database/service_db.go

```diff
@@ -52,59 +75,98 @@ func (d *LevelDBServiceDB) unmarshal(id string, value []byte) (*service.Service,
 func (d *LevelDBServiceDB) All() ([]*service.Service, error) {
 	var (
 		services []*service.Service
-		iter     = d.db.NewIterator(nil, nil)
+		iter     = d.db.NewIterator(util.BytesPrefix([]byte(idKeyPrefix)), nil)
 	)
-
 	for iter.Next() {
-		s, err := d.unmarshal(string(iter.Key()), iter.Value())
+		id := strings.TrimPrefix(string(iter.Key()), idKeyPrefix)
+		s, err := d.unmarshal(id, iter.Value())
 		if err != nil {
 			// NOTE: Ignore all decode errors (possibly due to a service
 			// structure change or database corruption)
 			if decodeErr, ok := err.(*DecodeError); ok {
 				logrus.WithField("service", decodeErr.ID).Warning(decodeErr.Error())
 				continue
 			}
+			iter.Release()
 			return nil, err
 		}
 		services = append(services, s)
 	}
 	iter.Release()
-	if err := iter.Error(); err != nil {
-		return nil, err
-	}
-
-	return services, nil
+	return services, iter.Error()
 }
 
 // Delete deletes service from database.
-func (d *LevelDBServiceDB) Delete(id string) error {
-	return d.db.Delete([]byte(id), nil)
+func (d *LevelDBServiceDB) Delete(idOrAlias string) error {
+	tx, err := d.db.OpenTransaction()
+	if err != nil {
+		return err
+	}
+
+	s, err := d.get(tx, idOrAlias)
+	if err != nil {
+		tx.Discard()
+		return err
+	}
+
+	batch := &leveldb.Batch{}
+	batch.Delete([]byte(idKeyPrefix + s.ID))
+	batch.Delete([]byte(aliasKeyPrefix + s.Alias))
+	if err := tx.Write(batch, nil); err != nil {
+		tx.Discard()
+		return err
+	}
+	return tx.Commit()
 }
 
 // Get retrives service from database.
-func (d *LevelDBServiceDB) Get(id string) (*service.Service, error) {
-	b, err := d.db.Get([]byte(id), nil)
+func (d *LevelDBServiceDB) Get(idOrAlias string) (*service.Service, error) {
+	return d.get(d.db, idOrAlias)
+}
+
+// get retrives service from database by using r reader.
+func (d *LevelDBServiceDB) get(r leveldb.Reader, idOrAlias string) (*service.Service, error) {
+	// check if key is an alias, if yes then save id.
+	id, err := r.Get([]byte(aliasKeyPrefix+idOrAlias), nil)
+	if err != nil && err != leveldb.ErrNotFound {
+		return nil, err
+	} else if err == nil {
+		idOrAlias = string(id)
+	}
+
+	b, err := r.Get([]byte(idKeyPrefix+idOrAlias), nil)
 	if err != nil {
 		if err == leveldb.ErrNotFound {
-			return nil, &ErrNotFound{ID: id}
+			return nil, &ErrNotFound{ID: idOrAlias}
 		}
 		return nil, err
 	}
-
-	return d.unmarshal(id, b)
+	return d.unmarshal(idOrAlias, b)
 }
 
 // Save stores service in database.
 func (d *LevelDBServiceDB) Save(s *service.Service) error {
+	// check service
 	if s.ID == "" {
-		return errors.New("database: can't save service without id")
+		return errCannotSaveWithoutID
+	}
+	if s.Alias == "" {
+		return errCannotSaveWithoutAlias
 	}
+	if len(s.ID) == len(s.Alias) {
+		return errAliasSameLen
+	}
+
+	// encode service
 	b, err := d.marshal(s)
 	if err != nil {
 		return err
 	}
 
-	return d.db.Put([]byte(s.ID), b, nil)
+	batch := &leveldb.Batch{}
```

Comment:
batch are atomic, you can check out source code

Reply:
All is atomic as it works on snapschot. Get need a little fix

Reply:
The batch doesn't do transaction all the the time 
> // If the batch size is larger than write buffer, it may justified to write
	// using transaction instead. Using transaction the batch will be written
	// into tables directly, skipping the journaling.

Also if we can rely on batches and we are in this condition why not using it for the `Delete` function ?

```go
batch.Delete([]byte(aliasKeyPrefix + s.Alias))
if err := tx.Write(batch, nil); err != nil {
```

Reply:
In delete function we need to get the service first , which isn't possible with batches.

And about batch, this docs only apply to a transaction buffer size, but the overall batch is atomic in leveldb (you can check it reading code in depth or read orgninal leveldb docs (as this one from golang leveldb is rather poor)

Reply:
I think we should use transaction. I read articles on batch and they said it's not really a transaction. Transactions give us more flexibility and more control. And also it lock the db from concurrent read and write.

Reply:
Batch seems to either relaying on transaction if batch is too big, otherwise, it write in a single write (and of course wait for the db to be unlocked)